src/lib/SaferNames/Syntax.hs

```diff
@@ -0,0 +1,585 @@
+  _              -> throw CompilerErr $ "Not a pair or unit: " ++ show ty
+
+-- ((...((ans & x{n}) & x{n-1})... & x2) & x1) -> (ans, [x1, ..., x{n}])
+fromLeftLeaningConsListTy :: MonadError Err m => Int -> Type n -> m (Type n, [Type n])
+fromLeftLeaningConsListTy depth initTy = go depth initTy []
+  where
+    go 0        ty xs = return (ty, reverse xs)
+    go remDepth ty xs = case ty of
+      PairTy lt rt -> go (remDepth - 1) lt (rt : xs)
+      _ -> throw CompilerErr $ "Not a pair: " ++ show xs
+
+fromConsList :: MonadError Err m => Atom n -> m [Atom n]
+fromConsList xs = case xs of
+  UnitVal        -> return []
+  PairVal x rest -> (x:) <$> fromConsList rest
+  _              -> throw CompilerErr $ "Not a pair or unit: " ++ show xs
+
+type BundleDesc = Int  -- length
+
+bundleFold :: a -> (a -> a -> a) -> [a] -> (a, BundleDesc)
+bundleFold empty pair els = case els of
+  []  -> (empty, 0)
+  [e] -> (e, 1)
+  h:t -> (pair h tb, td + 1)
+    where (tb, td) = bundleFold empty pair t
+
+mkBundleTy :: [Type n] -> (Type n, BundleDesc)
+mkBundleTy = bundleFold UnitTy PairTy
+
+mkBundle :: [Atom n] -> (Atom n, BundleDesc)
+mkBundle = bundleFold UnitVal PairVal
+
+pattern FunTy :: Binder n l -> EffectRow l -> Type l -> Type n
+pattern FunTy b eff bodyTy = Pi (Abs b (WithArrow (PlainArrow eff) bodyTy))
+
+pattern PiTy :: Binder n l -> Arrow l -> Type l -> Type n
+pattern PiTy b arr bodyTy = Pi (Abs b (WithArrow arr bodyTy))
+
+pattern BinaryFunTy :: Binder n l -> Binder l l' -> EffectRow l' -> Type l' -> Type n
+pattern BinaryFunTy b1 b2 eff bodyTy = FunTy b1 Pure (FunTy b2 eff bodyTy)
+
+pattern BinaryFunVal :: Binder n l -> Binder l l' -> EffectRow l' -> Block l' -> Atom n
+pattern BinaryFunVal b1 b2 eff body =
+          Lam (Abs b1 (WithArrow PureArrow (Block Empty (Atom (
+          Lam (Abs b2 (WithArrow (PlainArrow eff) body)))))))
+
+-- === instances ===
+
+-- right-biased, unlike the underlying Map
+instance Semigroup (SourceNameMap n) where
+  m1 <> m2 = SourceNameMap $ fromSourceNameMap m2 <> fromSourceNameMap m1
+
+instance Monoid (SourceNameMap n) where
+  mempty = SourceNameMap mempty
+
+instance Show (Block n) where
+  show _ = "TODO"
+
+instance Show (DataDef n) where
+  show _ = "TODO"
+
+-- variant with the common extension built in
+withTraverseCoreB
+  :: (SubstB b, Monad m)
+  =>                              Scope  o  -> CoreTraversal m  i  o  -> b i i'
+  -> (forall o'. FreshExt o o' -> Scope  o' -> CoreTraversal m  i' o' -> b o o' -> m a)
+  -> m a
+withTraverseCoreB s t b cont = do
+  traverseCoreB s t b >>= \case
+    FreshBinder ext b' renamer -> do
+      let t' = extendInjectNameTraversal ext renamer t
+      let s' = extendScope s b'
+      cont ext s' t' b'
+
+-- shorthand to makes instance implementations easier
+tce :: SubstE e => Monad m => Scope o -> CoreTraversal m i o -> e i -> m (e o)
+tce = traverseCoreE
+
+instance SubstB PlainBinder where
+  traverseCoreB = undefined
+  asBindingsFrag = undefined
```

Comment:
There are many undefined instances here. Is that a blocker?

Reply:
Follow-up PR!